File: codes/rust/72.编辑距离.rs
/*
 * @lc app=leetcode.cn id=72 lang=rust
 *
 * [72] 编辑距离
 */

// @lc code=start
impl Solution {
    pub fn min_distance(word1: String, word2: String) -> i32 {
        let m = word1.len();
        let n = word2.len();
        let mut dp: Vec<Vec<i32>> = vec![vec![0; n + 1]; m + 1];
        // 初始化base case
        for i in 0..=m {
            dp[i][0] = i as i32;
        }
        for j in 0..=n {
            dp[0][j] = j as i32;
        }
        // 开始状态转
        for i in 1..=m {
            for j in 1..=n {
                if word1.get(i - 1..i) == word2.get(j - 1..j) {
                    dp[i][j] = dp[i - 1][j - 1];
                } else {
                    let del_op = dp[i - 1][j] + 1; // 删除操作
                    let ins_op = dp[i][j - 1] + 1; // 插入操作
                    let rep_op = dp[i - 1][j - 1] + 1; // 替换操作
                    dp[i][j] = std::cmp::min(del_op, ins_op).min(rep_op);
                }
            }
        }
        dp[m][n]
    }
}
// @lc code=end

File: codes/golang/72.编辑距离.go
/*
 * @lc app=leetcode.cn id=72 lang=golang
 *
 * [72] 编辑距离
 */

/*
动态规划解法:
base case:  字符串1或字符串2走完，这个时候的base case是另外一个字符串剩下
的长度
状态的转移： 遍历s1的下标递增或s2的下标递增
方法： 自顶向下
状态变换方向： 以字符串1为例子，s1插入一个字符、删除一个字符、替换一个字符
*/

// @lc code=start
func minDistance(word1 string, word2 string) int {
	m := len(word1)
	n := len(word2)

	// 创建dp数组
	mp := make([][]int,m+1)
    mp[0] = make([]int,n+1)
	for i:=1; i < m+1; i++ {
		mp[i] = make([]int, n+1)
		mp[i][0] = i
	}
	for j:=1; j < n+1; j++ {
		mp[0][j] = j
	}

	// 自顶向下求解
	for i:=1; i <=m; i++ {
		for j:=1; j<=n; j++ {
			if word1[i-1] == word2[j-1] {
				mp[i][j] = mp[i-1][j-1]
			} else {
				mp[i][j] = mins(
					mp[i-1][j-1] + 1,
					mp[i-1][j] + 1,
					mp[i][j-1] +1,
				)
			}
		}
	} 
	return mp[m][n]
}

func mins(a,b,c int) int {
	return min(a,min(b,c))
}

func min(a,b int) int {
	if a > b {
		return b
	}
	return a
}

// @lc code=end

